### PR TITLE
[WIP] Persistent Kernel Examples - Old

### PR DESCRIPTION
**Pull request by rdspring1**
Related  #389 

- [x] Create Persistent Softmax Kernel
- [x] Create Persistent Batch Norm Kernel

**Issue:**
- Disabled [reorderGroupBreadthFirst](https://github.com/csarofeen/pytorch/blob/20_8_18_devel/torch/csrc/jit/codegen/cuda/lower_loops.cpp#L639-L640) because it incorrectly ordered the kernels operations

### CODE REVIEW

**Comment by rdspring1** [comment]
**Softmax Persistent**
**Generated Cuda Kernel:**
```
__global__ void CUDAGeneratedKernel(Tensor<float, 2> T0, Tensor<float, 2> T6) {
  alignas(4) extern __shared__ char array[];
  unsigned offset = 0;
  void* shared_mem = array;
  offset += ((blockDim.x * blockDim.y * blockDim.z) * sizeof(float));

  offset = alignBufferSize(offset,4);
  float* T4 = reinterpret_cast<float*>(array + offset);
  offset += (((ceilDiv(T0.size[1], 128)) * 128) * sizeof(float));

  float T3[1];

  float T2[1];
  T2[0] = float(0);

  float T7[1];
  T7[0] = float(0);

  // Read Input into Shared Memory
  for(size_t i4 = 0; i4 < (ceilDiv(T0.size[1], 128)); ++i4) {
    if ((((i4 * 128) + threadIdx.x) < T0.size[1])) {
      T4[((i4 * 128) + threadIdx.x)]
         = expf(T0[(blockIdx.x * T0.stride[0]) + (((i4 * 128) + threadIdx.x) * T0.stride[1])]);
    }
  }
  __syncthreads();

  // Reduction
  for(size_t i11 = 0; i11 < (ceilDiv(T0.size[1], 128)); ++i11) {
    float T1[1];
    if ((((i11 * 128) + threadIdx.x) < T0.size[1])) {
      T1[0]
         = T4[((i11 * 128) + threadIdx.x)];
    }
    if ((((i11 * 128) + threadIdx.x) < T0.size[1])) {
      T7[0]
        = T7[0]
        + T1[0];
    }
  }
  // Block Reduction
  blockReduce<true, false, false>(
    T2[0],
    T7[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    1,
    float(0));

  // Block Broadcast
  broadcast::blockBroadcast<true, false, false>(
    T3[0],
    T2[0],
    static_cast<float*>(shared_mem));

  // Division + Read from shared memory to registers
  for(size_t i32 = 0; i32 < (ceilDiv(T0.size[1], 128)); ++i32) {
    float T5[1];
    if ((((i32 * 128) + threadIdx.x) < T0.size[1])) {
      T5[0]
         = T4[((i32 * 128) + threadIdx.x)];
    }
    if ((((i32 * 128) + threadIdx.x) < T0.size[1])) {
      T6[(blockIdx.x * T6.stride[0]) + ((i32 * 128) + threadIdx.x)]
        = T5[0]
        / T3[0];
    }
  }
}
```

**Comment by rdspring1** [comment]
**BatchNorm**
**Generated Cuda Kernel:**
```
__global__ void CUDAGeneratedKernel(Tensor<float, 2> T0, float f4, float f5, float f6, int64_t i7, Tensor<float, 2> T17) {
  alignas(4) extern __shared__ char array[];
  unsigned offset = 0;

  void* shared_mem = array;
  offset += ((blockDim.x * blockDim.y * blockDim.z) * sizeof(float));

  offset = alignBufferSize(offset,4);
  float* T1 = reinterpret_cast<float*>(array + offset);
  offset += (((ceilDiv(T0.size[1], 128)) * 128) * sizeof(float));

  float T12[1];
  float T11[1];
  float T10[1];
  float T9[1];

  float T8[1];
  T8[0] = float(0);
  float T19[1];
  T19[0] = float(0);

  float T5[1];
  float T3[1];
  float T2[1];
  
  T2[0] = float(0);
  float T18[1];
  T18[0] = float(0);

  // Read into shared memory
  for(size_t i4 = 0; i4 < (ceilDiv(T0.size[1], 128)); ++i4) {
    if ((((i4 * 128) + threadIdx.x) < T0.size[1])) {
      T1[((i4 * 128) + threadIdx.x)]
         = T0[(blockIdx.x * T0.stride[0]) + (((i4 * 128) + threadIdx.x) * T0.stride[1])];
    }
  }
  __syncthreads();

  // Thread Independent Sum
  for(size_t i11 = 0; i11 < (ceilDiv(T0.size[1], 128)); ++i11) {
    if ((((i11 * 128) + threadIdx.x) < T0.size[1])) {
      T18[0]
        = T18[0]
        + T1[((i11 * 128) + threadIdx.x)];
    }
  }
  // Block Reduction - Mean
  blockReduce<true, false, false>(
    T2[0],
    T18[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    1,
    float(0));

  // Block Broadcast
  broadcast::blockBroadcast<true, false, false>(
    T3[0],
    T2[0],
    static_cast<float*>(shared_mem));

  // Mean
  T5[0]
    = T3[0]
    / i7;

  for(size_t i35 = 0; i35 < (ceilDiv(T0.size[1], 128)); ++i35) {
    float T7[1];
    float T6[1];
    float T4[1];
    // Load data from shared memory
    if ((((i35 * 128) + threadIdx.x) < T0.size[1])) {
      T4[0]
         = T1[((i35 * 128) + threadIdx.x)];
    }
    // Subtract mean
    if ((((i35 * 128) + threadIdx.x) < T0.size[1])) {
      T6[0]
        = T4[0]
        - T5[0];
    }
    // Square difference
    if ((((i35 * 128) + threadIdx.x) < T0.size[1])) {
      T7[0] = powf(
        T6[0],
        float(2));
    }
    // Thread Independent Sum
    if ((((i35 * 128) + threadIdx.x) < T0.size[1])) {
      T19[0]
        = T19[0]
        + T7[0];
    }
  }

  // Block Reduction - Variance
  blockReduce<true, false, false>(
    T8[0],
    T19[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    1,
    float(0));

  // Block Broadcast
  broadcast::blockBroadcast<true, false, false>(
    T9[0],
    T8[0],
    static_cast<float*>(shared_mem));

  // Variance
  T10[0]
    = T9[0]
    / i7;

  // Var + EPS
  T11[0]
    = T10[0]
    + f6;

  // 1 / Sqrt(Var + EPS)
  T12[0]
     = rsqrtf(T11[0]);

  for(size_t i90 = 0; i90 < (ceilDiv(T0.size[1], 128)); ++i90) {
    float T16[1];
    float T15[1];
    float T14[1];
    float T13[1];

    // Read from shared memory
    if ((((i90 * 128) + threadIdx.x) < T0.size[1])) {
      T13[0]
         = T1[((i90 * 128) + threadIdx.x)];
    }

    // Subtract mean
    if ((((i90 * 128) + threadIdx.x) < T0.size[1])) {
      T14[0]
        = T13[0]
        - T5[0];
    }

    // Norm = (x - mean) / sqrt(var + eps)
    if ((((i90 * 128) + threadIdx.x) < T0.size[1])) {
      T15[0]
        = T14[0]
        * T12[0];
    }

    // Norm * Gamma
    if ((((i90 * 128) + threadIdx.x) < T0.size[1])) {
      T16[0]
        = T15[0]
        * f4;
    }

    // Norm * Gamma + Beta
    if ((((i90 * 128) + threadIdx.x) < T0.size[1])) {
      T17[(blockIdx.x * T17.stride[0]) + ((i90 * 128) + threadIdx.x)]
        = T16[0]
        + f5;
    }
  }
}
```
